Packs/SecurityIntelligenceServicesFeed/Integrations/SecurityIntelligenceServicesFeed/SecurityIntelligenceServicesFeed.py

```diff
@@ -337,35 +369,78 @@ def indicator_field_mapping(feed_type: str, indicator: Dict[str, Any]) -> Dict[s
-                    fields['sisexpiration'] = prepare_date_string_for_custom_fields(
-                        indicator['MaliciousExpiration'])
-            else:
-                if indicator.get('Category'):
-                    fields['siscategory'] = indicator['Category']
-
-                if indicator.get('Expiration'):
-                    fields['sisexpiration'] = prepare_date_string_for_custom_fields(indicator['Expiration'])
-
-    except ValueError as exception:
-        raise ValueError(MESSAGES['Feed_EXTRACT_ERROR'] + str(exception))
+            if indicator.get('Category'):
+                fields['siscategory'] = indicator['Category']
+
+            if indicator.get('Expiration'):
+                fields['sisexpiration'] = prepare_date_string_for_custom_fields(indicator['Expiration'])
+    remove_nulls_from_dictionary(fields)
     return fields
 
 
+def validate_first_fetch_interval(first_fetch_interval):
+    """
+    Validating first fetch interval. it should be in form of (<number> <time unit>, e.g., 12 hours, 7 days,
+    3 months, 1 year)
+    raise value error if validation fails.
+
+    :param first_fetch_interval: first fetch interval
+    :return: None
+    """
+    range_split = first_fetch_interval.split(' ')
+    if len(range_split) != 2:
+        raise ValueError(MESSAGES['INVALID_FIRST_FETCH_INTERVAL_ERROR'])
+    if not range_split[0].isdigit():
+        raise ValueError(MESSAGES['INVALID_FIRST_FETCH_INTERVAL_ERROR'])
+    if not range_split[1] in ['minute', 'minutes', 'hours', 'hour', 'day', 'days', 'month', 'months', 'year', 'years']:
+        raise ValueError(MESSAGES['INVALID_FIRST_FETCH_UNIT_ERROR'])
+
+
+def get_latest_key(client: Client, feed_type: str, first_fetch_interval: str,
+                   is_get_indicators: bool = False) -> Tuple[str, int]:
+    """
+    Get latest key from bucket of specified feed type.
+    :return:
+    """
+    # Retrieving cached key from integration context.
+    cached_key, start_from = get_last_key_from_integration_context(feed_type)
+    object_key_list = client.request_list_objects(feed_type=feed_type, start_after=cached_key,
+                                                  prefix=feed_type.lower())
+    object_key_list = [object_key for object_key in object_key_list if '.gz' in object_key.get('Key', '')]
+    object_key_list.sort(key=lambda key_dict: key_dict['LastModified'])
+
+    if is_get_indicators:
+        return object_key_list[-1].get('Key', '') if object_key_list else cached_key, 0
+
+    # Parsing first fetch time.
+    date_from, now = dateparser.parse(f'{first_fetch_interval} UTC'), datetime.now(timezone.utc)
+
+    # Fetching latest object keys.
+    latest_key_list: List[str] = [key_dict.get('Key', '') for key_dict in
+                                  object_key_list if
+                                  key_dict.get('LastModified', now) >= date_from]
+
+    return latest_key_list[0] if latest_key_list and not start_from else cached_key, start_from
```

Comment:
Are we assuming `latest_key_list` will never be an empty list? it doesn't seem to be a correct assumption to me, am I missing something?

Reply:
@hod-alpert , If `latest_key_list` is empty, it will return cached_key which contains `key` or `''`. In fetch-indicators we have check if latest_key found then go ahead.